Publish tasks to workflow specific queues

prismofeverything opened the pull request:
This PR creates the ability to send tasks to specific queues for each workflow. Actually, it doesn't even mention queues because that's not how rabbit works as it turns out, when publishing you just provide the exchange and the routing key, and it gets routed to a queue that is bound to that routing key in that exchange. This required some changes in Sisyphus which is reflected in a version bump for Sisyphus in `project.clj`

In addition, I expanded the launch functionality to include the ability to specify metadata in both `launch-sisyphus.sh` and from the gaia python client in `client/python/gaia/client.py`. I have verified the right metadata fields get set on launch.

1fish2 commented on src/gaia/sisyphus.clj:
```diff
@@ -58,6 +62,14 @@
+   compute project zone
+   (if (empty? filters)
+     {}
+     {:filters filters})))
```
Did you mean `:filter`?

Move this function next to `cloud/list-instances`?

1fish2 commented on src/gaia/sisyphus.clj:
```diff
@@ -68,6 +80,13 @@
 (defn load-sisyphus-executor
-  "required keys are :rabbit and :kafka"
+  "Required keys are :rabbit and :kafka. Replaces configuration under :rabbit key
+  with rabbitmq connection map from `sisyphus.rabbit`."
```
I don't understand this docstring.

1fish2 commented on client/python/gaia/client.py:
```diff
@@ -47,11 +47,22 @@ def pop_path(path):
+            metadata_fields.append('{}={}'.format(key, value))
+        launch_metadata = ','.join(metadata_fields)
+
+    os.system("{} {} {}".format(command, worker, launch_metadata))
```
On further reflection, let's replace `launch-sisyphus.sh` and `launch-workers.sh` with a Python script that uses `subprocess32.call()` to call `gcloud`, provides a Python API for the workflow builder and others to call, and provides an argparse CLI for interactive debugging.

That'll avoid bash arrays and quoting hell while making it straightforward to pass in metadata fields.

It can get the app-specific parameters (machine-type, service-account, scopes, image-family...) from **kwargs or from an app-specific file.

prismofeverything replied:
> The worker-launching code in wcEcoli's runscripts/cloud/util/workflow.py + runscripts/cloud/launch-workers.sh moved beyond gaia/client

Ahaha okay, no worries I could have known that as well. I think launching is a gaia/sisyphus functionality so might have to migrate that back. 

Alternatively, I could see if we can use this compute service feature to make the launch call directly from the JVM code? And not have to muck around with shelling out to scripts etc. Python script could work also, whatever works!

prismofeverything commented on client/python/gaia/client.py:
```diff
@@ -121,12 +132,20 @@ def expire(self, workflow, keys):
         assert isinstance(names, list), 'need a list of worker names'
+        if metadata:
+            metadata = {
+                'rabbit-' + key: value
```
Good point, clearly I was hacking things in here.

prismofeverything commented on src/gaia/sisyphus.clj:
```diff
@@ -47,7 +48,10 @@
   (let [command (get commands (keyword (:command step)))
         task (step->task step command)]
-    (rabbit/publish! rabbit task)
+    (log/debug! "TASK: " task)
```
Sounds good to me : ) I'll stand down there.

prismofeverything commented on resources/config/gaia.clj:
```diff
@@ -13,7 +13,10 @@
-  :path ""}
+  :path ""
+  :rabbit {}
+  :project "gcloud-project"
```
Awesome yes, I'll make that change.

prismofeverything replied:
Okay, 

```
(log/shell-out "gcloud" "config" "get-value" "core/project")
```

works but

```
(log/shell-out "gcloud" "config" "get-value" "compute/zone")
```

does not. Doing a `gcloud config list` shows no config fields related to zone. So it looks like we are at a combination of HTTP metadata calls and shelling out to `gcloud`? Still better than hard-coding it in a config file though I suppose.

1fish2 commented:
I like this idea! So we'll implement worker launching today as the builder
et al sending a request to the Gaia server which would in turn use the Java
Compute API to launch the VMs. Later (days or weeks) we'll move the
initiative from the client side into Gaia.
    Jerry

On Thu, Nov 21, 2019, 12:42 PM Ryan Spangler <notifications@github.com>
wrote:

> *@prismofeverything* commented on this pull request.
> ------------------------------
>
> In client/python/gaia/client.py
> <https://github.com/prismofeverything/gaia/pull/2#discussion_r349307803>:
>
> >      command = os.path.join("script", "launch-sisyphus.sh")
>      if not os.path.exists(command):
>          command = "launch-sisyphus.sh"
> -    os.system("{} {}".format(command, key))
> +
> +    worker = options.get('worker', 'sisyphus')
> +    metadata = options.get('metadata', {})
> +
> +    launch_metadata = ''
> +    if metadata:
> +        metadata_fields = []
> +        for key, value in metadata.items():
> +            metadata_fields.append('{}={}'.format(key, value))
> +        launch_metadata = ','.join(metadata_fields)
> +
> +    os.system("{} {} {}".format(command, worker, launch_metadata))
>
> The worker-launching code in wcEcoli's runscripts/cloud/util/workflow.py +
> runscripts/cloud/launch-workers.sh moved beyond gaia/client
>
> Ahaha okay, no worries I could have known that as well. I think launching
> is a gaia/sisyphus functionality so might have to migrate that back.
>
> Alternatively, I could see if we can use this compute service feature to
> make the launch call directly from the JVM code? And not have to muck
> around with shelling out to scripts etc. Python script could work also,
> whatever works!
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/prismofeverything/gaia/pull/2?email_source=notifications&email_token=AAH6VMCKT5G74QDHRSNDA7TQU3XERA5CNFSM4JP3OS62YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCMSLTFY#discussion_r349307803>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAH6VMFPGZBEDSUJEUVVG3DQU3XERANCNFSM4JP3OS6Q>
> .
>

prismofeverything commented:
>  So we'll implement worker launching today as the builder et al sending a request to the Gaia server which would in turn use the Java Compute API to launch the VMs.

I'll see if it works!

1fish2 commented:
To fix that, run this:

gcloud config set compute/zone us-west1-b


There's some note about that in docs/update-sisyphus-server.md

It's time to update the sisyphus server disk image anyway.
    *Jerry*


On Tue, Dec 3, 2019 at 12:41 AM Ryan Spangler <notifications@github.com>
wrote:

> *@prismofeverything* commented on this pull request.
> ------------------------------
>
> In resources/config/gaia.clj
> <https://github.com/prismofeverything/gaia/pull/2#discussion_r353043445>:
>
> > @@ -13,7 +13,10 @@
>
>   :executor
>   {:target "sisyphus"
> -  :path ""}
> +  :path ""
> +  :rabbit {}
> +  :project "gcloud-project"
>
> Okay,
>
> (log/shell-out "gcloud" "config" "get-value" "core/project")
>
> works but
>
> (log/shell-out "gcloud" "config" "get-value" "compute/zone")
>
> does not. Doing a gcloud config list shows no config fields related to
> zone. So it looks like we are at a combination of HTTP metadata calls and
> shelling out to gcloud? Still better than hard-coding it in a config file
> though I suppose.
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/prismofeverything/gaia/pull/2?email_source=notifications&email_token=AAH6VMEHR43A2BGUJOJK2HTQWYLTJA5CNFSM4JP3OS62YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCNW6BFA#discussion_r353043445>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAH6VMD5EBE55BPR6MMGFQ3QWYLTJANCNFSM4JP3OS6Q>
> .
>

1fish2 commented:
sisyphus@sisyphus-jerry-demo-20191202-234412-0:~/sisyphus$ gcloud config
set compute/zone us-west1-b
Updated property [compute/zone].
sisyphus@sisyphus-jerry-demo-20191202-234412-0:~/sisyphus$ lein repl
nREPL server started on port 36651 on host 127.0.0.1 - nrepl://
127.0.0.1:36651
REPL-y 0.4.3, nREPL 0.6.0
Clojure 1.10.0
OpenJDK 64-Bit Server VM 11.0.4+11-post-Ubuntu-1ubuntu218.04.3
    Docs: (doc function-name-here)
          (find-doc "part-of-name-here")
  Source: (source function-name-here)
 Javadoc: (javadoc java-object-or-class-here)
    Exit: Control+D or (exit) or (quit)
 Results: Stored in vars *1, *2, *3, an exception in *e

sisyphus.core=> (require '[sisyphus.task :as task])
nil
sisyphus.core=> (task/shell-out "gcloud" "config" "get-value"
"core/project")
"allen-discovery-center-mcovert"
sisyphus.core=> (task/shell-out "gcloud" "config" "get-value"
"compute/zone")
"us-west1-b"

prismofeverything commented:
Okay I think I've addressed all feedback, and @1fish2 I merged in your PR and have tested this locally now and everything is working. I also pushed sisyphus 0.0.23 to clojars. Want to merge and test on GCE?

prismofeverything commented:
Actually, it's going to need the changes from https://github.com/CovertLab/sisyphus/pull/27 so maybe check those out first.

1fish2 commented:
Please do go for it. And the Gaia client to PYPI. (Heh, [sisyphus on PYPI](https://pypi.org/project/sisyphus/).)

If you need the latest client changes for this, they're in the branch https://github.com/CovertLab/wcEcoli/tree/workflow-properties . (I'll open the PR and do more testing after running out now to do some errands and then doing Travis' code review.)

prismofeverything commented:
Alright, merging and pushing to PYPI! Thanks Jerry. 

(good find on that sisyphus lib, convergent evolution!)

prismofeverything commented:
Gaia 0.0.9 is up on PYPI.
